text/0502_aws-vpclattice.md

```diff
@@ -0,0 +1,506 @@
+
+
+```typescript
+/**
+ * Create a vpc lattice service network.
+ * Implemented by `ServiceNetwork`.
+ */
+e/**
+ * Create a vpc lattice service network.
+ * Implemented by `ServiceNetwork`.
+ */
+export interface IServiceNetwork extends core.IResource {
+
+  /**
+  * The Amazon Resource Name (ARN) of the service network.
+  */
+  readonly serviceNetworkArn: string;
+
+  /**
+   * The Id of the Service Network
+   */
+  readonly serviceNetworkId: string;
+  /**
+   * Grant Princopals access to the Service Network
+   */
+  grantAccessToServiceNetwork(principal: iam.IPrincipal[]): void;
+  /**
+   * Add Lattice Service Policy
+   */
+  addService(service: Service): void;
+  /**
+   * Associate a VPC with the Service Network
+   */
+  associateVPC(props: AssociateVPCProps): void;
+  /**
+   * Log To S3
+   */
+  logToS3(bucket: s3.Bucket | s3.IBucket ): void;
+  /**
+   * Send Events to Cloud Watch
+   */
+  sendToCloudWatch(log: logs.LogGroup | logs.ILogGroup ): void;
+  /**
+   * Stream to Kinesis
+   */
+  streamToKinesis(stream: kinesis.Stream | kinesis.IStream ): void;
+  /**
+   * Share the ServiceNetwork
+   */
+  share(props: ShareServiceNetworkProps): void;
+  /**
+   * Create and Add an auth policy to the Service Network
+   */
+  applyAuthPolicyToServiceNetwork(): void;
+}
+
+}
+``` 
+All of the class functions, apart from addAuthPolicy, are convience functions that mirror the properties. 
+
+`ServiceNetwork` will take `ServiceNetworkProps` as props
+```typescript
+/**
+ * The properties for the ServiceNetwork.
+ */
+export interface ServiceNetworkProps {
+
+  /** The name of the Service Network. If not provided Cloudformation will provide
+   * a name
+   * @default cloudformation generated name
+   */
+  readonly name?: string
+
+  /** The type of  authentication to use with the Service Network.
+   * @default 'AWS_IAM'
+   */
+  readonly authType?: AuthType | undefined;
+
+  /**
+   * S3 buckets for access logs
+   * @default no s3 logging
+   */
+
+  readonly s3LogDestination: s3.IBucket[] | undefined;
+  /**
+   * Cloudwatch Logs
+   * @default no logging to cloudwatch
+   */
+  readonly cloudwatchLogs?: logs.ILogGroup[] | undefined;
+
+  /**
+   * kinesis streams
+   * @default no streaming to Kinesis
+   */
+  readonly kinesisStreams?: kinesis.IStream[];
+
+  /**
+   * Lattice Services that are assocaited with this Service Network
+   * @default no services are associated with the service network
+   */
+  readonly services?: Service[] | undefined;
+
+  /**
+   * Vpcs that are associated with this Service Network
+   * @default no vpcs are associated
+   */
+  readonly vpcs?: ec2.IVpc[] | undefined;
+
+  /**
+   * Account principals that are permitted to use this service
+   * @default none
+   */
+  readonly accounts?: iam.AccountPrincipal[] | undefined;
+
+  /**
+   * arnToShareWith, use this for specifying Orgs and OU's
+   * @default false
+   */
+  readonly arnToShareServiceWith?: string[] | undefined;
```

Comment:
It would be the arns of orgs, and our OU's..     I was'nt entirely happy with this..  I'm good to combine them if its appropropriate ( along with the accounts )..    This results in two things happening;

(a) It will share the service network, and it will create auth policy that will permit those principals to use them.

Reply:
More thinking about this.  This actually provides the end user with a way to create them self lots of badness, and bad practice.    Best practice is not to share the the servicenetwork, but the services that are attached to them.     I will however replace this with a couple of methods, so if they really insist, they still can..     .share() and .addStatementToAuthPolicy().

.addStatement should only *really* be used to add actions which have DENY effects.     Think of it a bit like a 'SCP' for the Service network.    however, users will be users and they really might have a legit reason, so we shoud'nt entirely block them.    We could perhaps add a warning to this.

Reply:
I've added two new optional properties to the ServiceNetwork;

allowExternalPrincipals
allowUnauthentictedAcces:  

These will set/unset statements in the Auth Policy, and will default to the 'secure' setting.